x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/task/gap_auto_fill_scheduler_task.ts

```diff
@@ -0,0 +1,417 @@
+                end: now.toISOString(),
+                sortOrder,
+                hasUnfilledIntervals: true,
+                ruleTypes: config.ruleTypes,
+              });
+
+              if (!ruleIds.length) {
+                await logEvent({
+                  status: GAP_AUTO_FILL_STATUS.SKIPPED,
+                  results: [],
+                  message: 'Skipped execution: no rules with gaps',
+                });
+                return { state: {} };
+              }
+
+              const aggregatedByRule = new Map<
+                string,
+                {
+                  ruleId: string;
+                  processedGaps: number;
+                  status: GapFillSchedulePerRuleStatus;
+                  error?: string;
+                }
+              >();
+
+              // Step 4: Process rules in batches
+              for (
+                let startIdx = 0;
+                startIdx < ruleIds.length;
+                startIdx += DEFAULT_RULES_BATCH_SIZE
+              ) {
+                // Stop early if we've reached capacity
+                if (remainingBackfills <= 0) {
+                  const consolidated = resultsFromMap(aggregatedByRule);
+                  await logEvent({
+                    status: GAP_AUTO_FILL_STATUS.SUCCESS,
+                    results: consolidated,
+                    message: `Stopped early: no backfill capacity remaining | ${formatConsolidatedSummary(
+                      consolidated
+                    )}`,
+                  });
+                  return { state: {} };
+                }
+                if (
+                  await handleCancellation({
+                    abortController,
+                    aggregatedByRule,
+                    logEvent,
+                  })
+                ) {
+                  return { state: {} };
+                }
+
+                const currentRuleIds = ruleIds.slice(startIdx, startIdx + DEFAULT_RULES_BATCH_SIZE);
+
+                // we need to find the rules that are enabled and
+                // and check for rule types and consumers that are supported
+                const { data: rules } = await rulesClient.find({
+                  options: {
+                    page: 1,
+                    perPage: currentRuleIds.length,
+                    filter: `alert.attributes.enabled:true AND (${currentRuleIds
+                      .map((id) => `alert.id: ("alert:${id}")`)
+                      .join(' OR ')})`,
+                  },
+                });
+
+                const toProcessRuleIds = rules.map((rule) => rule.id).slice(0, remainingBackfills);
+
+                if (!toProcessRuleIds.length) {
+                  continue;
+                }
+
+                let searchAfter: SortResults[] | undefined;
+                let pitId: string | undefined;
+
+                const gapsPerPage = DEFAULT_GAPS_PER_PAGE;
+                let gapFetchIterationCount = 0;
+
+                while (true) {
```

Comment:
I wonder can `nextSearchAfter`  be used as an indicator when we still need to process gaps?
```ts

while (nextSearchAfter) {
...
}
```

Once it's undefined, we terminate the loop.
Otherwise we have constructions like
```ts

          if (!filteredGaps.length) {
                    if (gapsPage.length < gapsPerPage) {
                      break;
                    }
                    continue;
                  }
```

which are difficult to read